### ISSUE TITLE
Scalability

### ISSUE THREAD

**Issue by slavakurilyak** [produvia/ai-platform#31]
**Goals**

As a developer, I want to figure out how to manage 100s or 1,000s of AI tasks, so that I can scale the ai-platform better.

As a developer, I want to experiment and create 10s or 100s of variations of each AI task, so that I can determine the most performant machine learning algorithm and method for solving each task. 

**Motivation**

We currently store each AI task into a separate folder (see [/tasks](https://github.com/produvia/ai-platform/tree/master/tasks) folder). This allows us to develop independent machine learning models and conduct experiments on each task using MLflow. The current folder structure also aligns with [MLflow storage format](https://www.mlflow.org/docs/latest/models.html#storage-format). As the ai-platform repository grows in size, we need to better scale the management of independent AI tasks.

**Possible Solutions**

1. Use distributed storage, such as: [AWS S3](https://aws.amazon.com/s3/), [DBFS](https://docs.databricks.com/user-guide/dbfs-databricks-file-system.html), or [Google Drive](https://www.google.com/drive/), to minimize input (models or datasets) and output (models). 

1. Use submodules, such as: [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) or [Google Repo](https://code.google.com/archive/p/git-repo/), to manage multi­-repository integrations in Git.

2. Use virtual filesystems, such as: [VFS for Git](https://vfsforgit.org/) (developed by Microsoft) or [Git LFS](https://git-lfs.github.com/) (developed by Atlassian), to enable git to operate at enterprise-scale. Microsoft wrote about it's motivations to to launch VFS for Git: [here](https://www.infoq.com/news/2017/02/GVFS/) and [here](https://docs.microsoft.com/en-us/azure/devops/learn/git/gvfs-design-history). Atlassian wrote a tutorial about Git LFS: [here](https://www.atlassian.com/git/tutorials/git-lfs).

**Additional Considerations**

MLflow was designed to scale to large data sets, large output files (for example, models), and large numbers of experiments ([MLflow Docs, 2019](https://www.mlflow.org/docs/latest/concepts.html#scalability-and-big-data)).

> 1. An individual MLflow run can execute on a distributed cluster, for example, using Apache Spark. You can launch runs on the distributed infrastructure of your choice and report results to a Tracking Server to compare them. MLflow includes a built-in API to launch runs on Databricks.
> 2. MLflow supports launching multiple runs in parallel with different parameters, for example, for hyperparameter tuning. You can simply use the Projects API to start multiple runs and the Tracking API to track them.
> 3. MLflow Projects can take input from, and write output to, distributed storage systems such as AWS S3 and DBFS. MLflow can automatically download such files locally for projects that can only run on local files, or give the project a distributed storage URI if it supports that. This means that you can write projects that build large datasets, such as featurizing a 100 TB file.

Let's consider dataset sizes used for training machine learning models. Imagine ai-platform contains 100 models and each dataset is 100 GB (i.e. image dataset). Storing datasets on ai-platform will require 10 TB. If ai-platform contains 1,000 models, that translates to 100 TB.

Let's consider machine learning model sizes used to solve tasks. Imagine ai-platform contains 100 models and each model is 100 MB. Storing models on ai-platform will require 10 GB. If ai-platform contains 1,000 models, that translates to 100 GB.

**Comment by slavakurilyak** [comment]
@lucky-singh, @mlaradji I would love to hear your thoughts on this matter

**Comment by lucky-singh** [comment]
hi @slavakurilyak, thank you for raising this issue at an early stage. 

Currently I didn't find contribution where we are training models and almost every case pre trained models are used, which are provided by the original researchers. However as soon as we would get custom use cases, we would be using real power of MLflow and train and save multiple models.

When we experiment with multiple hyper parameters then as described by you the size will explode and I thought it would not be ok to keep results into repository and I have put experiments path in .gitignore so that repository can be restricted to code base only.

Currently I was thinking for only management of pre trained models and example files used as an input. Multiple model results would be in present in consumer's local repository and they would be free to use MLflow recommendations on saving, tracking and sharing of models.

I read more about sub modules and in majority of the cases I came across word "complexity". However in case of ai-platform I guess it's a good solution because of following reasons:
1. Unlike other repositories we don't have inter dependencies with other part of the repository and considering our consumers would be interested in one use case / task at a time then they probably wouldn't want to wait for download and navigate in rest parts.
2. MLflow has a method to run MLflow project using git repository URL. So if all task are also present as an individual repository then this would also be a nice use case where our consumers would directly start exploring models without downloading repository locally.
3. it's an open source project and if we use only github, our contributors would be able to contribute without any storage cost involved. Otherwise one might choose to delete data from source and functionality might break in future. (With this reason, I would also recommend to have all repository within Produvia organization, so that we can have a assurance of standardization and consistent behavior.)

**Comment by slavakurilyak** [comment]
@lucky-singh you share some excellent points! A few ideas to consider:

1. The AI community is developing pre-trained models to save on development time and deal with shortage of training data. 
2. Computer vision and natural language processing (NLP) are two largest communities driving adoption of pre-trained models. Google recently described it's motivation for pre-training Tensorflow models to solve NLP tasks ([Google AI Blog, 2018](https://ai.googleblog.com/2018/11/open-sourcing-bert-state-of-art-pre.html)). 
3. As of today, there are at least 1,100+ AI tasks ([PapersWithCode, 2019](http://paperswithcode.com/sota)) which can be tracked and managed using **ai-platform**.
4. Ignoring model results using `.gitignore` is a great idea.
5. Once we implement AutoML (Issue #7), each task will have hundreds or thousands of variations.
5. We need to figure out how to deal with training data. If necessary, we can force all contributors to upload all model input (aka datasets) onto distributed storage platforms, such as AWS S3. If we do this, we must also force all contributors to include license (i.e. commercial or non-commercial) which describes the legal rights to each dataset.
6. Imagine there are 10,000 models in the wild today and 100,000 models next year. If **ai-platform** is able to manage these models, we can become the go-to source for model development and management.
7. Once we implement meta-learning (Issue #32), and/or automation of search and discovery (Issue #33), **ai-platform** can become a platform with managerial components that delegates and communicates with pools of AI tasks. (Issue #34)

**Comment by slavakurilyak** [comment]
@lucky-singh, @mlaradji What are your thoughts on using [Git Large File System (LFS)](https://git-lfs.github.com/)? This git extension for versioning large files allows us to store files up to 2 GB in size ([Github, 2019](https://help.github.com/en/articles/about-git-large-file-storage)). This could be useful when allowing developers to upload pre-built models.

**Comment by lucky-singh** [comment]
hi @slavakurilyak, Yes LFS could be a great solution or alternative of submodules.

I was thinking to have another git repository for pre-trained models. This way models can be downloaded from there. If contributors needs a pre trained model and unable to identify a place than keeping with tasks then they could request an upload to alternate repository.

Otherwise I am concerned about the time taken to clone the repository. Recently one pull request took 2-3 minutes in cloning and this time can only increase.

**Comment by slavakurilyak** [comment]
@lucky-singh separating models into its own repository is an idea worth exploring.

Produvia's **ai-platform** resembles a *model zoo* - a collection of machine learning models.

Let's consider TensorFlow as a case study. [TensorFlow](](https://github.com/tensorflow/tensorflow)) was first released on [November 6, 2015](https://github.com/tensorflow/tensorflow/commit/f41959ccb2d9d4c722fe8fc3351401d53bcf4900). The TensorFlow team realized that pre-trained models will take up a significant space on their repository. As the result, Tensorflow Team released [models](https://github.com/tensorflow/models) repository on [Jan 12, 2016](https://github.com/tensorflow/models/commit/15f82d209f1428d948d0467a573741c15f607e43). 

Did the TensorFlow Team store their pre-trained models on their [models](https://github.com/tensorflow/models) repo? No, they only stored code examples. All of their pre-trained models and datasets are now hosted on [http://download.tensorflow.org](https://github.com/tensorflow/models/search?q=http%3A%2F%2Fdownload.tensorflow.org&unscoped_q=http%3A%2F%2Fdownload.tensorflow.org). Datasets are hosted on http://download.tensorflow.org/data/ and pre-trained models are hosted on http://download.tensorflow.org/models/.

**Comment by slavakurilyak** [comment]
We need to figure out how to merge several machine learning models into one. @mlaradji originally brought up the idea [here](https://github.com/produvia/ai-platform/pull/72#pullrequestreview-270509945). 

Consider merging all [image classification pull requests](https://github.com/produvia/ai-platform/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+image+classification) into one. How would that look like?

I'm open to ideas and discussion around this topic.

**Comment by slavakurilyak** [comment]
As of today, there are 1,223+ AI tasks ([PapersWithCode, 2019](https://paperswithcode.com/sota)) which we can implement on our repository. Our AI Platform currently solves four tasks ([AI Platform, 2019](https://github.com/produvia/ai-platform/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+label%3Atask)).
